### ISSUE TITLE
Project 2 Level 3 - Random Generator Approach

### ISSUE THREAD

**Issue by ZhangAnli** [nus-cs2030/2021-s1#312]
# Description 
Hello, my approach for the Random Generator is to pass the Random Generator(RNG) along from Arrive Event -> (Optional) Wait Event -> Serve Event. Then when the RNG reaches Serve Event, I will use the .genServiceTime() method within the Serve Event. However, I have noticed that the question says: 
<img width="1120" alt="Screenshot 2020-11-11 at 10 25 06 AM" src="https://user-images.githubusercontent.com/65236091/98757768-2db35d80-2408-11eb-9b1a-7b0c65c9c78a.png">

Can I check with you guys if it is alright to just pass the RNG in its raw, unwrapped form in this manner? Or is it better to pass it as a supplier like () -> randomGenerator, and then call .get when I reach the Serve Event? This is because there is hint of a lazy approach and I don't think passing along the RNG in its raw form is Lazy Implementation? But it seems kind of pointless to me to wrap the RNG in a supplier then .get it only when it reaches the Serve Event? 

Thanks in advance!

**Comment by SkyBlaise99** [comment]
The question is asking you to parse `() -> rng.genServiceTime()`, but it is perfectly fine to just parse `rng` as a whole and let the `Event` determine when to call the gen methods.

**Comment by weiyifei-evelyn** [comment]
Hi! I have the same query as you, and I managed to understand why in level 3, the question hints the use of Lazy.
If you just pass the RNG without wrapping a supplier, it will pass all the test cases in level 3. It is because at Level 3, the test cases given are just nice that nth customer to arrive is just the nth customer to be served.

However, if you use unwrapped RNG in level 4, the test output will contain some error. So it is recommended to adopt a Lazy approach at your level 3. This can be done simply by making Service Time `() -> RNG.genServiceTime()`

Hope this helps you!

**Comment by bwangpj** [comment]
> Hi! I have the same query as you, and I managed to understand why in level 3, the question hints the use of Lazy.
> If you just pass the RNG without wrapping a supplier, it will pass all the test cases in level 3. It is because at Level 3, the test cases given are just nice that nth customer to arrive is just the nth customer to be served.
> 
> However, if you use unwrapped RNG in level 4, the test output will contain some error. So it is recommended to adopt a Lazy approach at your level 3. This can be done simply by making Service Time `() -> RNG.genServiceTime()`
> 
> Hope this helps you!

Correct me if I'm wrong, but from a theoretical standpoint there should be no difference between supplier and the actual rng in terms of handling whether the nth customer to arrive is the nth customer to be served. It all depends on when you call `genServiceTime`.

**Comment by weiyifei-evelyn** [comment]
> > Hi! I have the same query as you, and I managed to understand why in level 3, the question hints the use of Lazy.
> > If you just pass the RNG without wrapping a supplier, it will pass all the test cases in level 3. It is because at Level 3, the test cases given are just nice that nth customer to arrive is just the nth customer to be served.
> > However, if you use unwrapped RNG in level 4, the test output will contain some error. So it is recommended to adopt a Lazy approach at your level 3. This can be done simply by making Service Time `() -> RNG.genServiceTime()`
> > Hope this helps you!
> 
> Correct me if I'm wrong, but from a theoretical standpoint there should be no difference between supplier and the actual rng in terms of handling whether the nth customer to arrive is the nth customer to be served. It all depends on when you call `genServiceTime`.

Yes, that's right. In this case, making it Lazy ensures that `genServiceTime` is only executed when a person is being served. This is achieved by calling `.get()` on the supplier.

**Comment by SkyBlaise99** [comment]
@weiyifei-evelyn You do not have to make it `Lazy` to ensures that `genServiceTime` is only executed when a person is being served. There are other approach, and parsing the whole `RNG` is one of them. Using `Lazy` is also one of the approach, but not the only way.

**Comment by weiyifei-evelyn** [comment]
> @weiyifei-evelyn You do not have to make it `Lazy` to ensures that `genServiceTime` is only executed when a person is being served. There are other approach, and parsing the whole `RNG` is one of them. Using `Lazy` is also one of the approach, but not the only way.

Ohhh okay, I think I get what you mean! Both are correct approach.
